joint_space_mdp.py

```diff
@@ -0,0 +1,482 @@
+                    # expected_future_cost = 0
+
+                total_value = value + self.gamma*expected_future_value
+                sim_log.append((ns, value, expected_future_value, total_value))
+
+                if self.cost_based:
+                    found_better = (total_value < best_value)
+                else:
+                    found_better = (total_value > best_value)
+                if len(best_actions) == 0 or found_better:
+                    best_value = total_value
+                    best_actions = [ai]
+                elif math.isclose(total_value, best_value, abs_tol=1e-6):
+                    best_actions.append(ai)
+
+            # self.plot_sim_results(sim_log)
+            if self.DEBUG:
+                filename = "logs/action_costs_of_%d_%d_iter_%d" % (
+                    xi, yi, self.main_iter)
+                np.save(filename, sim_log)
+
+            self.P[yi][xi] = best_actions
+            end = time.time()
+            print("Time(s) for one state: %.3f" % (end - start))
+            total_time += (end-start)
+        print("Total Runtime of Update Policy: %.3f" % total_time)
+
+    def init_action_space(self, run_full_mdp):
+        A = []  # need to maintain order
+        self.da = math.pi/80
+        if run_full_mdp:
+            self.dh = 5
+            self.velocities = np.arange(start=0.1, stop=0.31, step=0.1)
+            self.angle_offsets = np.arange(
+                start=-2*self.da, stop=3*self.da, step=self.da)
+        else:
+            self.dh = 3
+            self.velocities = np.arange(start=0.1, stop=0.21, step=0.05)
+            self.angle_offsets = np.arange(
+                start=-2*self.da, stop=3*self.da, step=self.da)
+
+        self.contact_heights = np.arange(
+            start=self.env.bottle.height/self.dh,
+            stop=self.env.bottle.height + self.env.bottle.height/self.dh,
+            step=self.env.bottle.height/self.dh)
+
+        self.dr = 0.25  # proportion of max reach
+        self.reach_ranges = np.arange(
+            start=self.dr,
+            stop=1.0+self.dr,
+            step=self.dr)
+
+        for h in self.contact_heights:
+            for v in self.velocities:
+                for a in self.angle_offsets:
+                    for r in self.reach_ranges:
+                        action = Action(
+                            angle_offset=a, velocity=v, height=h, reach_p=r)
+                        A.append(action)
+
+        return A
+
+    def plot_sim_results(self, sim_log):
+        # n = len(sim_log)
+        # color_vals = np.random.randint(0, 0xFFFFFF, size=n)  # +1 for target
+        # colors = [('#%06X' % v) for v in color_vals]
+        for (_, start, end) in sim_log:
+            print(start, end)
+            plt.plot([start[0], end[0]], [start[1], end[1]])
+
+        plt.legend([str(action) for action in self.A], loc='upper left')
+        plt.show()
+
+    def test_reach_p(self):
```

Comment:
More good places for unit tests! (Btw if you want advice on how best to do unit tests in python, let me know.)